assembly/std/index.ts

```diff
@@ -102,6 +103,103 @@ export function balanceOf(address: string): u64 {
   return env.balanceOf(address);
 }
 
+/**
+ * Check for key in datastore
+ *
+ * @param {StaticArray<u8>} key
+ *
+ * @return {bool} - true if key is present in datastore, false otherwise.
+ */
+export function hasOpKey(key: Uint8Array): bool {
+  let result = env.hasOpKey(key.buffer);
+  // From https://doc.rust-lang.org/reference/types/boolean.html &&
+  // https://www.assemblyscript.org/types.html
+  // we can safely cast from u8 to bool
+  return bool(result[0]);
+}
+
+/*
+ * Get data associated with the given key from datastore
+ *
+ * @param {StaticArray<u8>} key
+ *
+ * @return {StaticArray<u8>} - data as a byte array
+ */
+export function getOpData(key: Uint8Array): Uint8Array {
+  let result_ = env.getOpData(key.buffer);
+  return Uint8Array.wrap(result_);
+}
+
+/*
+ * Get all keys from datastore
+ *
+ * @return {Array<StaticArray<u8>} - a list of key (e.g. a list of bytearray)
+ */
+export function getOpKeys(): Array<Uint8Array> {
+  let buf = env.getOpKeys();
+  let keys_ser: Uint8Array = Uint8Array.wrap(buf);
+  return derOpKeys(keys_ser);
+}
+
+/*
+ * Internal function - used by getOpKeys
+ */
+export function derOpKeys(keys_ser: Uint8Array): Array<Uint8Array> {
```

Comment:
I still believe the implem of derOpKeys is a bit too complicated. Input from the ABI towards the smart contract can be trusted to minimize gas usage and code complexity. The other way around is not true however: data sent from smart contracts to the rust abi should be checked because it is non-validated user input.


Here is a recommendation (also avoids creating default_res when it is not needed):

```
export function derOpKeys(keys_ser: Uint8Array): Array<Uint8Array> {
  if (keys_ser.length == 0) {
    return new Array<Uint8Array>();
  }

  // Datastore key list deserialization
  // Format is: L (u32); V1_L (u8); V1 data (u8*V1_L); ...
  // u8 * 4 (LE) => u32
  let dv = new DataView(keys_ser.buffer, 0, 4);
  let entry_count = dv.getUint32(0, true /* littleEndian */);
  
  let cursor = 4;
  let keys_der = new Array<Uint8Array>(entry_count);
  for (let i = 0; i < entry_count; i++) {
    let end = cursor + keys_ser[cursor] + 1;
    keys_der[entry_pushed] = keys_ser.subarray(cursor + 1, end); // no copy here
    cursor = end;
  }
    
  return keys_der;
}
```

Reply:
Ok for trusting input from the ABI but this is not bullet proof. Any unchecked (or without enough review) modification (e.g from a newcomer) can cause an issue in deserialization. Moreover because we remove the checks, reporting the proper problem will not be possible.

Anyway it's done and unit tests about these edge cases & invalid format have been removed.